### PR TITLE
Resubmit "Fix possible in-use table after DETACH"

### PR DESCRIPTION
**Pull request by tavplubix**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Resubmit https://github.com/ClickHouse/ClickHouse/pull/45493
Reverted in https://github.com/ClickHouse/ClickHouse/pull/45545

See also https://github.com/ClickHouse/ClickHouse/issues/45554
Logs from the integration test:  https://pastila.nl/?027b12b7/685e1abd17ab38d39b3f9d407a32b0ea 

cc: @azat

### CODE REVIEW

File: src/Databases/DatabaseOnDisk.cpp [hunk 1]
```diff
@@ -379,14 +379,18 @@ void DatabaseOnDisk::renameTable(
...
         throw Exception(ErrorCodes::INCORRECT_QUERY, "Use RENAME/EXCHANGE TABLE (instead of RENAME/EXCHANGE DICTIONARY) for tables");
 
+    /// We have to lock the table before detaching, because otherwise lockExclusively will throw. But the table may not exist.
+    bool need_lock = table != nullptr;
+    if (need_lock)
+        table_lock = table->lockExclusively(local_context->getCurrentQueryId(), local_context->getSettingsRef().lock_acquire_timeout);
+
     detachTable(local_context, table_name);
+    if (!need_lock)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Table was detached without locking, it's a bug");
```
Comment (antonio2368) [inline src/Databases/DatabaseOnDisk.cpp:389]: we expect it to throw?
Why don't we return before `detachTable` if table is nullptr?

Comment (tavplubix) [reply src/Databases/DatabaseOnDisk.cpp:389]: Yes, we expect `detachTable` to throw a correct exception

**Review by antonio2368 — APPROVED** [review]
I think the failure occurred almost always so if it doesn't happen in the current check it should be good indicator it's good now

**Comment by tavplubix** [comment]
Integration tests (tsan) [4/6] - https://github.com/ClickHouse/ClickHouse/issues/45435
Stateless tests (release, s3 storage) [2/2] - https://github.com/ClickHouse/ClickHouse/pull/45604
Stress test (ubsan) - `reportLogStats` failed due to memory limit, will fix in https://github.com/ClickHouse/ClickHouse/pull/45527

**Pull request merged**